File: 2020/java/src/main/java/eu/pineman/chall/aoc2020/One.java
package eu.pineman.chall.aoc2020;

import org.javatuples.Pair;
import org.javatuples.Triplet;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class One {
    // https://phauer.com/2019/modern-best-practices-testing-java/#dont-use-static-access-never-ever
    public static long[] getInput() throws IOException {
        return Utils.getInput(1).mapToLong(Long::parseLong).toArray();
    }

    public static long[] getBigBoy() throws IOException {
        return Utils.getBigBoy(1).mapToLong(Long::parseLong).toArray();
    }

    public static Pair<Long, Long> twoSum(long[] input, long target) {
        Arrays.sort(input);
        int l = 0, r = input.length - 1;
        while (r != l) {
            long sum = input[l] + input[r];
            if (sum == target) return Pair.with(input[r], input[l]);
            else if (sum < target) l++;
            else r--;
        }
        return Pair.with(-1L, -1L);
    }

    public static Triplet<Long, Long, Long> threeSum(long[] input, long target) {
        Arrays.sort(input);
        int i = 0;
        for (long a : input) {
            List<Long> slice = Arrays.stream(input).boxed().collect(Collectors.toList());
            slice.remove(i);
            Pair<Long, Long> ts = twoSum(slice.stream().mapToLong(s -> s).toArray(), target - a);
            if (a + ts.getValue0() + ts.getValue1() == target) {
                return Triplet.with(a, ts.getValue0(), ts.getValue1());
            }
            i++;
        }
        return Triplet.with(-1L, -1L, -1L);
    }

    public static long partOne(long[] input, long target) {
        Pair<Long, Long> sol = twoSum(input, target);
        return sol.getValue0() * sol.getValue1();
    }

    public static long partTwo(long[] input, long target) {
        Triplet<Long, Long, Long> sol = threeSum(input, target);
        return sol.getValue0() * sol.getValue1() * sol.getValue2();
    }

    public static void main(String[] args) throws IOException {
        long[] input = getInput();
        System.out.println(partOne(input, 2020));
        System.out.println(partTwo(input, 2020));

        long[] bigBoy = getBigBoy();
        System.out.println(partOne(bigBoy, 99920044));
        System.out.println(partTwo(bigBoy, 99920044));
    }
}

File: 2020/go/aoc2020_one/one.go
package main

import (
	"errors"
	"fmt"
	"sort"
	"strconv"

	aoc2020 "github.com/pineman/code/chall/aoc2020/go"
)

func getInput() []int {
	return intInput(aoc2020.GetInput(1))
}

func getBigBoy() []int {
	return intInput(aoc2020.GetBigBoy(1))
}

func intInput(input []string) []int {
	intInput := make([]int, len(input))
	for i, v := range input {
		v, err := strconv.Atoi(v)
		if err != nil {
			panic(err)
		}
		intInput[i] = v
	}
	return intInput
}

func twoSum(input []int, target int) (int, int, error) {
	sort.Ints(input)
	l, r := 0, len(input)-1
	for l != r {
		s := input[l] + input[r]
		if s == target {
			return input[l], input[r], nil
		} else if s < target {
			l++
		} else {
			r--
		}
	}
	return -1, -1, errors.New("twoSum: could not find two numbers in input that sum to target")
}

func threeSum(input []int, target int) (int, int, int, error) {
	sort.Ints(input)
	for i, v := range input {
		a, b, _ := twoSum(append(input[:i], input[i+1:]...), target-v)
		if a+b+v == target {
			return a, b, v, nil
		}
	}
	return -1, -1, -1, errors.New("threeSum: could not find three numbers in input that sum to target")
}

func partOne(input []int, target int) int {
	a, b, err := twoSum(input, target)
	if err != nil {
		panic(err)
	}
	return a * b
}

func partTwo(input []int, target int) int {
	a, b, c, err := threeSum(input, target)
	if err != nil {
		panic(err)
	}
	return a * b * c
}

func main() {
	input := getInput()
	fmt.Println(partOne(input, 2020))
	fmt.Println(partTwo(input, 2020))

	bigBoy := getBigBoy()
	fmt.Println(partOne(bigBoy, 99920044))
	// fmt.Println(partTwo(bigBoy, 99920044))
}
